# 準備運動

### Q1. 数列の和
1からNまでの整数和を求める関数calc_sumを作成せよ。

In [14]:
def calc_sum(n):
    result = 0
    for i in range(n):
        result += i + 1
    return result

# 1から10までの整数の和を求めてみる
print(calc_sum(10))

55


### Q2. 平方根
97の正の平方根をmathモジュールを使わずに力づくで求めよ。なお、求める解は、その二乗と97との差が0.001以内に収まるように求めよ。

In [2]:
root = 0

while True:
    delta = abs(97 - root ** 2)
    if delta < 0.001:
        break
    else:
        root += 0.0001
        
print(root)

9.848899999990385


### Q3. 文の表示
変数x, yを受け取り「xの価格はy円（税込）」という文字列を返す関数を作成せよ（[言語処理100本ノック2015 Q.7](http://www.cl.ecei.tohoku.ac.jp/nlp100/)改題）。

In [3]:
def show_price(x, y):
    text = "{product}の価格は{price}円（税抜）".format(product=x, price=y)
    return text

show_price("リンゴ", "97")

'リンゴの価格は97円（税抜）'

### Q4. 回文
文字列"Rats live on no evil star"の文字を逆に並べた文字列を表示せよ。

In [4]:
text = "Rats live on no evil star"

reversed_text = ""
for i in range(len(text)):
    reversed_text += text[len(text) - i - 1]
    
print(reversed_text)

rats live on no evil staR


In [5]:
# 別解1
chars = list(text)
result = ""
for i in range(len(chars)):
    result += chars.pop()
print(result)

rats live on no evil staR


In [6]:
# 別解2
chars = list(text)
chars.reverse()
result = "".join(chars)
print(result)

rats live on no evil staR


### Q5. 円周率
"Now I need a drink, alcoholic of course, after the heavy lectures involving quantum mechanics."という文を単語に分解し、各単語の（アルファベットの）文字数を先頭から出現順に並べたリストを作成せよ（[言語処理100本ノック 2015 Q.3](http://www.cl.ecei.tohoku.ac.jp/nlp100/)より）。

In [7]:
text = "Now I need a drink, alcoholic of course, after the heavy lectures involving quantum mechanics"

words = text.replace(",", "").split(" ")
lengths = []
for word in words:
    length = len(word)
    lengths.append(length)
print(lengths)

[3, 1, 4, 1, 5, 9, 2, 6, 5, 3, 5, 8, 9, 7, 9]


### Q6. 文字カウント
任意の文字列に含まれる文字の出現頻度を求める関数count_charを作成せよ。なお、返り値は、キーが文字、値が出現頻度に対応した辞書オブジェクト形式にせよ（例："I have a pen" -> {'I': 1, ' ': 3, 'h': 1, 'a': 2, 'v': 1, 'e': 2, 'p': 1, 'n': 1}）。

（ヒント）
ある文字列sが与えられたとき、list(s)はsを1文字ずつ分解したリストを返す。

In [8]:
products = ["apple", "orange", "melon"]
length = len(products)

for i in range(length):
    print(products[i])
    
for product in products:
    print(product)

apple
orange
melon
apple
orange
melon


In [9]:
def count_char(text):
    chars = list(text)
    frequency = {}
    for char in chars:
        frequency.setdefault(char, 0)
        frequency[char] += 1
    return frequency

text = input("任意の文字列を入力してください:")
frequency = count_char(text)
print(frequency)

任意の文字列を入力してください:任意の文字列を入力してください
{'任': 1, '意': 1, 'の': 1, '文': 1, '字': 1, '列': 1, 'を': 1, '入': 1, '力': 1, 'し': 1, 'て': 1, 'く': 1, 'だ': 1, 'さ': 1, 'い': 1}


### Q7. CSVファイルの読み込み
data/introduction_assignmentディレクトリにあるからpopulation-abstract2010-2015.csvをPythonから読み込み、その内容を一行ずつ表示せよ。

（留意事項）
当該CSVファイルは「[平成27年度国勢調査の総人口・世帯数](http://www.e-stat.go.jp/SG1/estat/GL08020103.do?_csvDownload_&fileId=000008040403&releaseCount=3)データ」の一部を抽出・加工したものです。


In [10]:
with open("../data/introduction_assignment/population-abstract2010-2015.csv") as cfile:
    for line in cfile:
        print(line.rstrip())

id,area_name,population2015,population2010,household2015,household2010
1000,北海道,5381733,5506419,2444810,2424317
2000,青森県,1308265,1373339,510945,513385
3000,岩手県,1279594,1330147,493049,483934
4000,宮城県,2333899,2348165,944720,901862
5000,秋田県,1023119,1085997,388560,390136
6000,山形県,1123891,1168924,393396,388608
7000,福島県,1914039,2029064,737598,720794
8000,茨城県,2916976,2969770,1124349,1088411
9000,栃木県,1974255,2007683,763097,745604
10000,群馬県,1973115,2008068,773952,755756
11000,埼玉県,7266534,7194556,2971659,2841595
12000,千葉県,6222666,6216289,2609132,2515904
13000,東京都,13515271,13159417,6701122,6393781
14000,神奈川県,9126214,9048302,3979278,3844512
15000,新潟県,2304264,2374450,848150,839039
16000,富山県,1066328,1093247,391171,383439
17000,石川県,1154008,1169788,453368,441170
18000,福井県,786740,806314,279687,275599
19000,山梨県,834930,863075,330976,327721
20000,長野県,2098804,2152449,807108,794461
21000,岐阜県,2031903,2080773,753212,737151
22000,静岡県,3700305,3765007,1429600,1399140
23000,愛知県,7483128,7410719,3063833,2933802
240

### Q8. CSVファイルからのデータの切り出し
Q7で用いたCSVファイルは、下記項目からなる人口統計データである。当該CSVデータから、都道府県名をキー、2015年調査時と2010年調査時の人口増減数を値とする辞書オブジェクトpopulationを作成せよ。
* id
* 都道府県名（area_name）
* 2015年度調査時の人口（population2015）
* 2010年度調査時の人口（population2010）
* 2015年度調査時の世帯数（household2015）
* 2015年度調査時の世帯数（household2010）

In [11]:
import csv

population = {}
with open('../data/introduction_assignment/population-abstract2010-2015.csv') as cfile:
    reader = csv.reader(cfile)
    header = next(reader)

    for row in reader:
        area_name = row[1]
        delta = int(row[3]) - int(row[2])
        population[area_name] = delta

print(population)        

{'北海道': 124686, '青森県': 65074, '岩手県': 50553, '宮城県': 14266, '秋田県': 62878, '山形県': 45033, '福島県': 115025, '茨城県': 52794, '栃木県': 33428, '群馬県': 34953, '埼玉県': -71978, '千葉県': -6377, '東京都': -355854, '神奈川県': -77912, '新潟県': 70186, '富山県': 26919, '石川県': 15780, '福井県': 19574, '山梨県': 28145, '長野県': 53645, '岐阜県': 48870, '静岡県': 64702, '愛知県': -72409, '三重県': 38859, '滋賀県': -2139, '京都府': 25739, '大阪府': 25776, '兵庫県': 53333, '奈良県': 36412, '和歌山県': 38619, '鳥取県': 15226, '島根県': 23045, '岡山県': 23751, '広島県': 16760, '山口県': 46609, '徳島県': 29758, '香川県': 19579, '愛媛県': 46231, '高知県': 36180, '福岡県': -29588, '佐賀県': 16956, '長崎県': 49592, '熊本県': 31256, '大分県': 30191, '宮崎県': 31164, '鹿児島県': 58065, '沖縄県': -40748}


### Q9. 辞書オブジェクトのソート
Q7で用いたCSVファイルを利用して、2010年度調査から2015年度調査の間で人口増減数が大きかった都道府県を増減数付で上位10件を表示せよ。表示は「30位: 京都府（25739人）」という形式で行うこと。

In [12]:
sorted_population = sorted(population.items(), key=lambda x: x[1], reverse=True)

rank = 1
for area_name, num in sorted_population:
    if rank > 10:
        break
    else:
        message = "{rank}位: {area_name}（{num}人）".format(rank=rank, area_name=area_name, num=num)
        print(message)
    rank += 1

1位: 北海道（124686人）
2位: 福島県（115025人）
3位: 新潟県（70186人）
4位: 青森県（65074人）
5位: 静岡県（64702人）
6位: 秋田県（62878人）
7位: 鹿児島県（58065人）
8位: 長野県（53645人）
9位: 兵庫県（53333人）
10位: 茨城県（52794人）


### Q10. ファイルの書き込み・保存
Q7,Q8,Q9で用いた2015年度の国勢調査データを用いて、都道府県名、各都道府県の2010年度調査から2015年度調査の間で人口増減数について、TSV形式でファイル保存せよ。

In [13]:
with open("../data/output/population-delta.tsv", "w+") as tfile:
    for area_name, delta in sorted_population:
        row = "{area_name}\t{delta}\n".format(area_name=area_name, delta=delta)
        tfile.write(row)